contrib/scalar/scalar.c

```diff
@@ -19,55 +19,98 @@ static int is_unattended(void) {
+	char *path_sep = find_last_dir_sep(buf->buf + offset);
+	strbuf_setlen(buf, path_sep ? path_sep - buf->buf : offset);
+}
+
+/**
+ * Given an absolute path at or below the enlistment root, find the
+ * enlistment root and working directory
+ */
+static int find_enlistment(struct strbuf *path,
+			   struct strbuf *enlistment_root)
+{
+	char *base_dir;
+	size_t offset = offset_1st_component(path->buf);
+	do {
+		const size_t len = path->len;
+
+		/* check if currently in enlistment root with src/ workdir */
+		strbuf_addstr(path, "/src/.git");
+		if (is_git_directory(path->buf)) {
+			strbuf_strip_suffix(path, "/.git");
+
+			if (enlistment_root) {
+				strbuf_addbuf(enlistment_root, path);
+				strbuf_parentdir(enlistment_root);
+			}
+
+			return 0;
+		}
+
+		/* reset to original path */
+		strbuf_setlen(path, len);
+
+		/* check if currently in workdir */
+		strbuf_addstr(path, "/.git");
+		if (is_git_directory(path->buf)) {
+			strbuf_setlen(path, len);
+
+			if (enlistment_root) {
+				strbuf_addbuf(enlistment_root, path);
+
+				base_dir = find_last_dir_sep(enlistment_root->buf + offset);
+				if (base_dir && !strcmp(base_dir + 1, "src"))
+					strbuf_setlen(enlistment_root,
+						base_dir - enlistment_root->buf);
+			}
+
+			return 0;
+		}
+
+		/* reset path to parent */
+		strbuf_setlen(path, len);
+		strbuf_parentdir(path);
+	} while (path->len > offset);
+
+	return -1;
+}
+
 static void setup_enlistment_directory(int argc, const char **argv,
 				       const char * const *usagestr,
-				       const struct option *options)
+				       const struct option *options,
+				       struct strbuf *enlistment_root)
 {
+	struct strbuf path = STRBUF_INIT;
+
 	if (startup_info->have_repository)
 		BUG("gitdir already set up?!?");
 
 	if (argc > 1)
 		usage_with_options(usagestr, options);
 
+	/* find the worktree, determine its corresponding root */
 	if (argc == 1) {
-		char *src = xstrfmt("%s/src", argv[0]);
-		const char *dir = is_directory(src) ? src : argv[0];
-
-		if (chdir(dir) < 0)
-			die_errno(_("could not switch to '%s'"), dir);
-
-		free(src);
-	} else {
-		/* find the worktree, and ensure that it is named `src` */
-		struct strbuf path = STRBUF_INIT;
-
-		if (strbuf_getcwd(&path) < 0)
-			die(_("need a working directory"));
-
-		for (;;) {
-			size_t len = path.len;
-
-			strbuf_addstr(&path, "/src/.git");
-			if (is_git_directory(path.buf)) {
-				strbuf_setlen(&path, len);
-				strbuf_addstr(&path, "/src");
-				if (chdir(path.buf) < 0)
-					die_errno(_("could not switch to '%s'"),
-						  path.buf);
-				strbuf_release(&path);
-				break;
-			}
-
-			while (len > 0 && !is_dir_sep(path.buf[--len]))
-				; /* keep looking for parent directory */
+		strbuf_add_absolute_path(&path, argv[0]);
+	} else if(strbuf_getcwd(&path) < 0) {
+		die(_("need a working directory"));
+	}
 
-			if (!len)
-				die(_("could not find enlistment root"));
+	strbuf_trim_trailing_dir_sep(&path);
+	if (!!find_enlistment(&path, enlistment_root))
```

Comment:
We do not need the `!!` here, as _any_ non-zero return value of `find_enlistment()` will make this condition be evaluated as `true`. The `!!` is only needed if we need to turn all non-zero values into `1`.

Reply:
I think the rule for regular functions whose name does not imply a Boolean return value is that `-1` indicates an error and `0` indicates success.

For example, `is_directory()` _would_ imply a Boolean to me, whereas `find_enlistment()` would suggest to me that we expect to find an enlistment, and have to error out if none is found. Therefore, I would tend to use the standard Unix rule "`0` == success" here.